sklearn/cluster/_bisect_k_means.py

```diff
@@ -0,0 +1,448 @@
+        Returns
+        -------
+        centers : ndarray of shape (n_clusters = 2, n_features)
+        """
+        n_samples = X.shape[0]
+        n_clusters = 2
+
+        if init_size is not None and init_size < n_samples:
+            init_indices = random_state.randint(0, n_samples, init_size)
+            X = X[init_indices]
+            x_squared_norms = x_squared_norms[init_indices]
+            n_samples = X.shape[0]
+
+        if isinstance(init, str) and init == 'k-means++':
+            centers, _ = _kmeans_plusplus(X, n_clusters,
+                                          x_squared_norms, random_state)
+        elif isinstance(init, str) and init == 'random':
+            seeds = random_state.permutation(n_samples)[:n_clusters]
+            centers = X[seeds]
+        elif hasattr(init, '__array__'):
+            centers = init
+        elif callable(init):
+            centers = init(X, n_clusters, random_state=random_state)
+            centers = check_array(centers,
+                                  dtype=X.dtype, copy=False, order='C')
+            self._validate_center_shape(X, centers)
+
+        if sp.issparse(centers):
+            centers = centers.toarray()
+
+        return centers
+
+    def _bisect(self, X, y=None, sample_weight=None, random_state=None):
+        """ Bisection of data
+        Attempts to get best bisection of data by performing regular K-Means
+        for different pairs of centroids
+
+        .. note:: Number of attempts is specified by self.n_init
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix} of shape (n_samples, n_features)
+
+            Training instances to cluster.
+
+            .. note:: The data will be converted to C ordering,
+                which will cause a memory copy
+                if the given data is not C-contiguous.
+
+        y : Ignored
+            Not used, present here for API consistency by convention.
+
+        sample_weight : array-like of shape (n_samples,), default=None
+            The weights for each observation in X. If None, all observations
+            are assigned equal weight.
+
+        Returns
+        -------
+        self
+            Fitted estimator.
+        """
+        X = self._validate_data(X, accept_sparse='csr',
+                                dtype=[np.float64, np.float32],
+                                order='C', copy=self.copy_x,
+                                accept_large_sparse=False)
+
+        # Validate init array
+        init = self.init
+
+        if hasattr(init, '__array__'):
+            init = check_array(init, dtype=X.dtype, copy=True, order='C')
+            self._validate_center_shape(X, init)
+
+        if not sp.issparse(X):
+            X_mean = X.mean(axis=0)
+            X -= X_mean
+
+            if hasattr(init, '__array__'):
+                init -= X_mean
+
+        x_squared_norms = row_norms(X, squared=True)
```

Comment:
Would it be possible to only compute that once, and not at each bisecting iteration ?

Reply:
I don't think so, since it is performed each time right after subtracting mean of remaining data , but I'll need to check that

Reply:
Subtracting the mean is just a trick for more accurate distance computations. We can do it once and for all and not necessarily at each bisect iteration.

Reply:
Good point